### PR TITLE
Image pull secrets

### LINKED ISSUE #257: add imagePullSecrets support
I'm receiving this error from Krustlet trying to pull images from a public (!) GCR registry.

As shown below, `wasm-to-oci` can push|pull images to this registry and I've configured it to support public (non-authenticated) access. I'm able to run regular containers from it on the cluster without problem.

```
kubectl describe pod/hello-world-wasi-golang
Name:         hello-world-wasi-golang
Namespace:    default
Priority:     0
Node:         krustlet/
Labels:       <none>
Annotations:  Status:  Failed
Message:      Failed to decode registry token from auth request
IP:           
IPs:          <none>
Containers:
  hello-world-wasi-golang:
    Image:      gcr.io/${PROJECT}/hello-world-wasi-golang:v1
    Port:       <none>
    Host Port:  <none>
    Environment:
      FOO:             bar
      POD_NAME:        hello-world-wasi-golang (v1:metadata.name)
      CONFIG_MAP_VAL:  <set to the key 'myval' of config map 'hello-world-wasi-golang'>  Optional: false
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-x75t7 (ro)
Conditions:
  Type           Status
  PodScheduled   True 
Volumes:
  default-token-x75t7:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-x75t7
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     krustlet/arch=wasm32-wasi:NoExecute
                 node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type    Reason     Age        From               Message
  ----    ------     ----       ----               -------
  Normal  Scheduled  <unknown>  default-scheduler  Successfully assigned default/hello-world-wasi-golang to krustlet
```
I'm able to pull regular containers from the same registry to the same cluster.

```bash
PROJECT=....

docker pull hello-world
docker tag hello-world gcr.io/${PROJECT}/hello-world
docker push gcr.io/${PROJECT}/hello-world

kubectl run hello-world --image=gcr.io/${PROJECT}/hello-world

kubectl logs pod/hello-world

Hello from Docker!
...
```

I'm able to `wasm-to-oci [push|pull]` images to the registry:

```bash
go run ./cmd push \
  hello-world-wasi-golang \
  gcr.io/${PROJECT}/hello-world-wasi-golang:v1
INFO[0004] Pushed: gcr.io/${PROJECT}/hello-world-wasi-golang:v1 
INFO[0004] Size: 2289810                                
INFO[0004] Digest: sha256:0e047d23...

go run ./cmd pull gcr.io/${PROJECT}/hello-world-wasi-golang:v1
INFO[0001] Pulled: gcr.io/${PROJECT}/hello-world-wasi-golang:v1 
INFO[0001] Size: 2289810                                
INFO[0001] Digest: sha256:0e047d23...

```
I tried retrieving a copy of `hello-world-wasi-rust:v0.1.0` from `webassembly.azurecr.io`, adding it to `gcr.io/${PROJECT}/hello-world-wasi-rust:v1` and then reconfiguring the Pod YAML to reference it, but I get the same error.
```
go run ./cmd pull webassembly.azurecr.io/hello-world-wasi-rust:v0.1.0
INFO[0001] Pulled: webassembly.azurecr.io/hello-world-wasi-rust:v0.1.0 
INFO[0001] Size: 1964621                                
INFO[0001] Digest: sha256:670adc71...

go run ./cmd push module.wasm gcr.io/${PROJECT}/hello-world-wasi-rust:v1
INFO[0003] Pushed: gcr.io/${PROJECT}/hello-world-wasi-rust:v1 
INFO[0003] Size: 1964621                                
INFO[0003] Digest: sha256:54b09224...
```

### PR DESCRIPTION
**Pull request by itowlson**
Fixes #257.

Do we have any thoughts on how to test it?  I can set up a private ACR registry, similar to `webassembly,azurecr.io`, and a SP with pull only permissions, but I'm a bit anxious about:

1. How do we keep track of/protect the registry so it doesn't accidentally get deleted in a tidy-up?
2. Posting a password, even a read-only one, in a public repo.

In the meantime I've tested it privately, and it seems okay though I'm worried that it seems to make the integration test flakes worse (and increasing the timeouts doesn't seem to help).

### CODE REVIEW

**Comment by thomastaylor312** [comment]
> Posting a password, even a read-only one, in a public repo

Can we just get away with it being a build secret set as an environment variable? We could have a "ci_only" feature for the tests. 

Let's talk offline about the test registry and see what we can come up with

**Comment by itowlson** [comment]
Simon suggested that and I was sceptical, but on reflection it might work.  Most code changes are unlikely to need the author to check pre-PR that they haven't broken image pull secrets - unless they are messing with the pull subsystem it should be transparent to them.  So having these tests run only in CI, as a backstop in case of unintended changes, would probably suffice.

File: crates/kubelet/src/secret/mod.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        let secret_futures: Vec<_> = self
+            .image_pull_secret_names
+            .iter()
+            .map(|name| secrets_api.get(name))
+            .collect();
+        let secret_results = futures::future::join_all(secret_futures).await;
+
+        for secret_result in secret_results {
+            match secret_result {
+                Err(e) => return Err(anyhow::Error::from(e)),
```
[the remark this answers is not in the document]
Comment (itowlson) [reply crates/kubelet/src/secret/mod.rs:43]: Ooh it does!  TIL - thanks!

**Comment by itowlson** [comment]
After discussion with @thomastaylor312 and @bacongobbler we are going to merge this as is, so that we can clear the decks for rebasing the state machine work onto the kubelet crate.  Work on testing (and any further iteration on this) can continue in separate PRs.

**Pull request merged**